service/src/builder.rs

```diff
@@ -80,209 +75,122 @@ fn extract_shuttle_toml_name(path: PathBuf) -> anyhow::Result<String> {
     let mut next_packages = Vec::new();
 
-    for member in ws.members() {
+    for member in metadata.workspace_packages() {
+        println!("{}", member.name);
         if is_next(member) {
             ensure_cdylib(member)?;
-            next_packages.push(member.name().to_string());
+            next_packages.push(member);
         } else if is_alpha(member) {
             ensure_binary(member)?;
-            alpha_packages.push(member.name().to_string());
+            alpha_packages.push(member);
         }
     }
 
     let mut runtimes = Vec::new();
 
     if !alpha_packages.is_empty() {
-        let opts = get_compile_options(&config, alpha_packages, release_mode, false)?;
-        let compilation = compile(&ws, &opts)?;
-
-        let mut alpha_binaries = compilation
-            .binaries
-            .iter()
-            .map(|binary| {
-                BuiltService::new(
-                    binary.path.clone(),
-                    false,
-                    binary.unit.pkg.name().to_string(),
-                    binary.unit.pkg.root().to_path_buf(),
-                    binary.unit.pkg.manifest_path().to_path_buf(),
-                )
-            })
-            .collect();
+        let mut compilation = compiler(alpha_packages, release_mode, false, project_path.clone())?;
+        trace!("alpha packages compiled");
 
-        runtimes.append(&mut alpha_binaries);
+        runtimes.append(&mut compilation);
     }
 
     if !next_packages.is_empty() {
-        let opts = get_compile_options(&config, next_packages, release_mode, true)?;
-        let compilation = compile(&ws, &opts)?;
+        let mut compilation = compiler(next_packages, release_mode, true, project_path)?;
+        trace!("next packages compiled");
 
-        let mut next_libraries = compilation
-            .cdylibs
-            .iter()
-            .map(|binary| {
-                BuiltService::new(
-                    binary.path.clone(),
-                    true,
-                    binary.unit.pkg.name().to_string(),
-                    binary.unit.pkg.root().to_path_buf(),
-                    binary.unit.pkg.manifest_path().to_path_buf(),
-                )
-            })
-            .collect();
-
-        runtimes.append(&mut next_libraries);
+        runtimes.append(&mut compilation);
     }
 
     Ok(runtimes)
 }
 
 pub fn clean_crate(project_path: &Path, release_mode: bool) -> anyhow::Result<Vec<String>> {
-    let (read, write) = pipe::pipe();
     let project_path = project_path.to_owned();
+    let manifest_path = project_path.join("Cargo.toml");
+    let mut profile = "dev";
+    if release_mode {
+        profile = "release";
+    }
+
+    // It is easier just to use several pipes
+    let (mut stderr_read, mut stderr_write) = pipe::pipe();
+    let (mut stdout_read, mut stdout_write) = pipe::pipe();
+    let (mut status_read, mut status_write) = pipe::pipe();
 
     tokio::task::spawn_blocking(move || {
```

Comment:
I think it might be blocking due to being used in the deployment handler. Say, if 10 people run this against production at the same time, it shouldn't block other async tasks.

Reply:
Hmm, at this moment the deployer is per project, so I think it would be more like 1 person made 10 requests against it. Even when this is the case, we usually queue these requests in the `deployer/deployment/queue.rs` logic, so I don't see why we would still need the `spawn_blocking` based on this.

In any case the current `spawn_blocking` is currently preceded by this comment:
```
// This needs to be on a separate thread, else deployer will block (reason currently unknown :D)
```

I think we still don't know why this needs to be blocking, but given we're working with a spawned `cargo` process now we should be able to start the process and then wait on its output asynchronously.